File: solutions/cpp/0030.cpp
class Solution {
 public:
  vector<int> findSubstring(string s, vector<string>& words) {
    if (s.empty() || words.empty()) return {};

    const int n = words[0].length();

    vector<int> ans;
    unordered_map<string, int> count;

    for (const string& word : words)
      ++count[word];

    for (int i = 0; i < s.length() - words.size() * n + 1; ++i) {
      unordered_map<string, int> seen;
      int j;
      for (j = 0; j < words.size(); ++j) {
        const string word = s.substr(i + j * n, n);
        if (++seen[word] > count[word]) break;
      }
      if (j == words.size())
        ans.push_back(i);
    }

    return ans;
  }
};

File: solutions/java/0030.java
class Solution {
  public List<Integer> findSubstring(String s, String[] words) {
    if (s.isEmpty() || words.length == 0)
      return new ArrayList<>();

    final int n = words[0].length();

    List<Integer> ans = new ArrayList<>();
    Map<String, Integer> count = new HashMap<>();

    for (String word : words)
      count.put(word, count.getOrDefault(word, 0) + 1);

    for (int i = 0; i <= s.length() - words.length * n; ++i) {
      Map<String, Integer> seen = new HashMap<>();
      int j = 0;
      for (; j < words.length; ++j) {
        String word = s.substring(i + j * n, i + j * n + n);
        seen.put(word, seen.getOrDefault(word, 0) + 1);
        if (seen.get(word) > count.getOrDefault(word, 0))
          break;
      }
      if (j == words.length)
        ans.add(i);
    }

    return ans;
  }
}